Implement lab05 stringer for anz-bank go-course

undrewb opened the pull request:
Fixes #525 

Review of colleague's PR #511 

#### Changes proposed in this PR:
- Implement lab05 stringer for anz-bank go-course
- Implement assertion based tests.

Commit:
Implement lab05 stringer for anz-bank go-course

Update test to fix newline issue

patrickmarabeas commented:
LGTM

As a side note - you can get rid of the `\n` in your `TestMainOutput` by using `fmt.Fprint` instead of `fmt.Fprintln` in `main()`

anzdaddy commented:
I advise against changing `main` just to make `TestMainOutput` slightly simpler. The use of `Fprintln` makes for cleaner output, e.g.:
```bash
user@machine:~$ lab05-fprintln
127.0.0.1
user@machine:~$ _
```

```bash
user@machine:~$ lab05-println
127.0.0.1user@machine:~$ _
```

anzdaddy commented on 05_stringer/undrewb/main_test.go:
```diff
@@ -0,0 +1,57 @@
+}{
+	{name: "localhost",
+		input: IPAddr{127, 0, 0, 1},
+		want:  "127.0.0.1"},
```
Cleaner:
```go
	{
		name: "localhost",
		input: IPAddr{127, 0, 0, 1},
		want:  "127.0.0.1",
	},
```

juliaogris replied:
yeah, this is strange. I'm surprised that `gofmt` doesn't complain.

juliaogris replied:
please fix.

undrewb replied:
fixed formatting of test case data

anzdaddy commented on 05_stringer/undrewb/main_test.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func TestStringer(t *testing.T) {
+	for _, tt := range stringerData {
```
Much cleaner:
```go
func TestIPAddrStringer(t *testing.T) {
	assertString := func(name string, expected string, input IPAddr) {
		t.Run(name, func(t *testing.T) {
			assert.Equal(t, expected, fmt.Sprint(input))
		})
	}

	assertString("localhost", "127.0.0.1", IPAddr{127, 0, 0, 1})
	assertString("four octet addr", "68.2.44.125", IPAddr{68, 2, 44, 125})
	⋮
}
```

juliaogris replied:
Disagree, @anzdaddy - @undrewb 's code is the typical table driven test look and feel.

undrewb replied:
leaving table driven test format

anzdaddy commented on 05_stringer/undrewb/main_test.go:
```diff
@@ -0,0 +1,57 @@
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			got := fmt.Sprint(tt.input)
+			assert.Equal(t, got, tt.want)
```
Incorrect parameter ordering. https://godoc.org/github.com/stretchr/testify/assert#Equal

juliaogris replied:
always in unit tests: first expected then actual. 
Please fix.

undrewb replied:
fixed parameter ordering to use expected then actual

anzdaddy commented on 05_stringer/undrewb/main_test.go:
```diff
@@ -0,0 +1,57 @@
+	main()
+
+	want := strconv.Quote("127.0.0.1\n")
+	got := strconv.Quote(buf.String())
```
Why `strconv.Quote`?

anzdaddy replied:
Actually, this seems to be a common pattern among go-course PRs. Is it recommended somewhere to do it this way?

juliaogris replied:
It saves you typing `"\"127.0.0.1\n\"" - which is the output of main, which needs to be tested for coverage. that's all.

juliaogris reviewed (changes requested):
Pretty close.

Commit:
Fix formatting issues with test cases raised by reviewers

Change assert parameter order to be expected then actual
Fix formatting of test case structs

rokane commented on 05_stringer/undrewb/main_test.go:
```diff
@@ -0,0 +1,68 @@
+		input: IPAddr{68, 2, 44, 125},
+		want:  "68.2.44.125",
+	},
+	{name: "three octet addr",
```
The indentation on this is inconsistent with the rest, can you please fix this so this line starts on the one below

rokane reviewed (changes requested):
Looking really good. I have just a real minor one about a single indentation case. If you can quickly fix that please ping me on slack, I should be able to verify and give this an approval afterwards.

Commit:
Fix indentation in test

Commit:
Fix formatting

juliaogris commented:
The go-course is now closed. Thank you very much for participating.
